### DISCUSSION TITLE
Automating app updates using PAC CLI

### DISCUSSION THREAD

**Discussion by parvezghumra** [microsoft/powerplatform-build-tools#722] [Ideas]
I'm trying to automate some of my environment maintenance activity. As part of this I was attempting to use the **pac application install** command, but this looks like it will not support **updating** of apps that are already installed but where an update is available in the target environment at runtime. It would be great if this would be made possible.

I found a previous issue that was logged and closed last year [here](https://github.com/microsoft/powerplatform-build-tools/issues/271#issuecomment-1905869275), but thought it would be worth starting a discussion as a new feature request

**Comment by jaehwanahn** [comment]
This needs more attention. It's not practical to go to each environment manually and apply the wave update while there are ~50 environments.

**Comment by AurelienClere-365** [comment]
Agree, should be great to have that in your radar/backlog please.

**Comment by jofme** [comment]
+1, this would be REALLY nice in our tenant 👍

**Comment by torbennehmer** [comment]
+1, full automation support is paramount

**Comment by TalviArt** [comment]
+1 for this. Does anyone know if it's currently possible to even list available updates of environments?

**Comment by macel94** [comment]
+1, we have 90+ environments and currently i do the clicks.
it's a pain :D

**Comment by lazejanev** [comment]
It should be of course through PAC CLI but also from the UI, there should be a multi select and update button that will update all of the selected apps

**Comment by TalviArt** [comment]
Heyy @petrochuk, any update on this? Is the ability to automate the app updates using 'pac application' or by other means on the roadmap? Thanks!

**Comment by laneswenka** [comment]
Hi everyone - we have added this to our backlog and will share when it is assigned to an engineer

**Reply by rnwood** [reply]
+1 and noting the need for this to work when authenticated as a Service Principal without the global PP/D365 admin role or flagging of the SP in some global allow list.

The current admin APIs rely on this and this makes it unsuitable for many customers where it being scoped by the environment SA role is needed. It's important this doesn't happen or this feature will be useless for automation.

**Reply by laneswenka** [reply]
Great point! This should rely on data plane permissions ( https://learn.microsoft.com/en-us/power-platform/admin/programmability-extensibility-overview#customer-management-plane-vs-customer-data-plane ) and not customer management plane.  That said, it would likely be a different issue for us to track if that is not working as expected.  If you have system customizer / system admin role in DV that should suffice for package import / update and solution import update.

**Comment by laneswenka** [comment]
This has been checked in and will be in the next monthly release of PAC CLI

**Reply by AurelienClere-365** [reply]
Hello Lane :)

Do you know the ETA for the next release of PAC CLI ? This will be the one of August ?
Also, for F&O part (as example), did you push arguments to be available for extra parameters needed when update is needed (like you know the version of F&O)

Thanks !
Aurelien

**Reply by miklos-molnar-annata** [reply]
Is this available already? I couldn't find a reference of this in the release notes.

**Reply by macel94** [reply]
@laneswenka news?

**Comment by petrochuk** [comment]
Closing this issue as done. Install command updates application. For example:

```
C:\> pac application install -env 39cf751c-*** --application-name mspcat_CatalogPackaging
Connected as ***
Verifying if application(s) are already installed...
'mspcat_CatalogPackaging' is already installed. Proceeding with update.
Installing 'mspcat_CatalogPackaging'...
Polling to check the status of your Application... Execution time: 00:00:30.1617529
Polling to check the status of your Application... Execution time: 00:01:00.3006289
Application successfully installed
Package Name            Operation ID
mspcat_CatalogPackaging 57ce5742-***
```

**Reply by naderisitc** [reply]
Hi @petrochuk, @laneswenka 
I tried installing the Dynamics 365 Finance and Operations Provisioning App app using _pac application install_ command but getting the following error with HttpStatusCode: 400. Any idea?

pac application install --environment 'xxxxxxxxx' --application-name 'msdyn_FinanceAndOperationsProvisioningAppAnchor'

_**"This application utilizes a Custom Install Experience through a Single Page Application. Install for such apps is not supported by this API. Please utilize the Power Platform Admin Center to initiate the install and configuration experience."**_

**Reply by AurelienClere-365** [reply]
I guess it's normal, the Provisioning  App for F&O in PPAC is using extra parameters/arguments. I guess this new update for PAC CLI is more update D365 Apps where it's not requiring any extra steps to do (should be the same as like D365 Marketing as example) or even Inventory Visibility which ask you a Client ID before installing. For me this new version is for basic apps where you just update in 2 click and helping saving a lot of time (like Dual Write and so on)

**Reply by AurelienClere-365** [reply]
But I agree like I said in a previous comment, if we can support adding extra arguments/parameters for update too, (like the version of F&O since we were talking about that) it will be great. But for me it's more opening a new issue or change requests.

**Reply by miklos-molnar-annata** [reply]
@laneswenka are there plans to enable F&O update through PAC CLI (or through other APIs)?

**Reply by miklos-molnar-annata** [reply]
> @laneswenka are there plans to enable F&O update through PAC CLI (or through other APIs)?
never mind, I have found the solution in [d365bap.tools](https://www.powershellgallery.com/packages/d365bap.tools/) :)
